### ISSUE TITLE
Benchmark Requirement for Synthetic Data Set

### ISSUE THREAD

**Issue by KevinBian107** [KevinBian107/GLearning-Benchmark#3]
# Considerata
The purpose of this issue is to ensure that everyone has the same or at least simiular of parameters and training status for benchmarking the synthetic data garph learning task.

**NOTE: THIS IS THE MINIMAL NEEDED, WE WILL GO UP FROM THIS**

### Training details

**General**:
- Optimizor: Adam
- Batch size: TBD
- Dropout: TBD
- Weight decay: TBD
- Learning rate: TBD 
- Dataset: start with ER

**Specific to Seq-Transformer**:
- Number of attention head: 1
- Positional encoding: idex based method

**Specific to Graph-Transformer**:
- Positional encoding: TBD

### Cycle Check
- Task: Binary graph-level classification.
- Loss: Logistic head + Cross Entropy.
- Performance Metric: Accuracy / confusion matrix.
- Efficiency Metric: Time per epoch / time per 1 percent acc increase / num of params (should be very similar).
- Graph (only for GPS and MPNN): give two example of the input native graph along with what the label (yes/no).

### Shortest Path
- Task: Multi-class node-level classification (the integer values are ordinal).
- Loss: Softmax head + MSE.
- Performance Metric: MSE / MAE / confusion matrix / recall / precision / accuracy / F1 scores.
- Efficiency Metric: Time per epoch / time per 1 percent acc increase / num of params.
- Graph (only for GPS and MPNN): give two example of the input native graph along with what the quert/target node and its length between label.

## Parameters Size
We will base our parameter size of [this "Understanding Transformer Reasoning Capabilities via Graph Algorithms" paper](https://arxiv.org/pdf/2405.18512). This is the minimal number of parameters that we need to achieve the task theoritically.

We are assuming:
- Number of nodes $N=32$ 
- Neglect the final MLP layer
- Assume using 1-attention head

### Tokenization + Seq-Transformer

Transformers will have 3 $(K, Q, V)$ matrix, each being $m x m$:
- Cycle check can be done by a transformer in **LogDepthPause** and **LogDepthWide** regimes, need $L = O(log N)$ depth and $m =O(N^e)$ where $e>0$ model size (model size does not grow linearly with the graph size).

- Shortest path can be done by a transformer in **LogDepthWide** regimes, need $L = O(log N)$ depth and $m = O(N^{0.5 + e})$ where $e>0$ model size (model size does not grow linearly with the graph size, but need more params).

Ignoring exponential constant, number of params for both task should be around $O(Lm^2) = O(log N x N^{2e})$:
- For a $N=32$ setup, this is $1024 x 5$ (depends on what log we are using, using log 5 now).
- This is roughly a $64 x 64$ 1 layer transformer block or a $32 x 32$ 4 layer transformer. However, this is before splitting for the three $(K, Q, V)$ matrix.
- After the split, this is roughly a 2 layer $32 x 32$ transformer block or a 7 layer $16 x 16$ transformer (hidden dimension = 16).

### MPNN
MPNN's weight matrix will have $d x k$ where $d$ is the dimension of node feature and $k$ is the model hidden size and (assuming no edge feature). $d$ is maximal 2 for us (shortest path task). We will have $2 x k$ number of param for one layer of MPNN. Since MPNN usually perform poorly in global task, we will over parametrize it to 3 - 4 layer of $k = 1028$.

### Graph GPS
Following same requirements as in the Tokenization + Seq-Transformer section. Since the GT and MPNN in GPS need the same size, we will use:
- Assuming MPNN in the front, then we can have roughly 1 layer of  GT with $32 x 32$ and 1 layer of MPNN with $2 x32$ (in total $32 x 32 x 3 + 2 x 32$) per GPS block.
- We can use 2 - 3 layers of the GPS block.
